crates/astria-conductor/src/celestia/reconstruct.rs

```diff
@@ -0,0 +1,201 @@
+                "sequencer header blob contains the target rollup ID, but no matching rollup blob was found; the blob",
+            );
+        } else {
+            reconstructed_blocks.push(ReconstructedBlock {
+                celestia_height,
+                block_hash: header_blob.block_hash(),
+                header: header_blob.into_unchecked().header,
+                transactions: vec![],
+            });
+        }
+    }
+    reconstructed_blocks
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+struct Key {
+    block_hash: [u8; 32],
+    height: SequencerHeight,
+}
+
+impl Key {
+    fn from_header_blob(val: &CelestiaSequencerBlob) -> Self {
+        Self {
+            block_hash: val.block_hash(),
+            height: val.height(),
+        }
+    }
+}
+
+struct BlockHashAndHeightToHeader {
+    // The sequencer header blobs grouped by (block_hash, sequencer_height)
+    map: HashMap<Key, CelestiaSequencerBlob>,
+
+    // A reverse index of block_hash -> [(block_hash, sequencer_height)]
+    reverse_index: HashMap<[u8; 32], Vec<Key>>,
+}
+
+impl BlockHashAndHeightToHeader {
+    fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+            reverse_index: HashMap::new(),
+        }
+    }
+
+    fn from_header_blobs(blobs: Vec<CelestiaSequencerBlob>) -> Self {
+        let mut this = Self::new();
+        for blob in blobs {
+            if let Some(dropped) = this.insert(blob) {
+                warn!(
+                    block_hash = %base64(&dropped.block_hash()),
+                    sequencer_height = %dropped.height(),
+                    "more than one sequencer header blob with the same block hash and height passed verification; dropping previous",
+                );
+            }
+        }
+        this
+    }
+
+    fn insert(&mut self, blob: CelestiaSequencerBlob) -> Option<CelestiaSequencerBlob> {
+        let key = Key::from_header_blob(&blob);
+        if let Some(old) = self.map.insert(key, blob) {
+            return Some(old);
+        }
+        self.reverse_index
+            .entry(key.block_hash)
+            .and_modify(|keys| keys.push(key))
+            .or_insert_with(|| vec![key]);
+        None
+    }
+
+    fn any_headers_with_block_hash(&self, block_hash: [u8; 32]) -> bool {
+        self.reverse_index.contains_key(&block_hash)
+    }
+
+    fn remove_header_blob_matching_rollup_blob(
+        &mut self,
+        rollup: &CelestiaRollupBlob,
+    ) -> Option<CelestiaSequencerBlob> {
+        let position_of_matching_key = self
+            .reverse_index
+            .get(&rollup.sequencer_block_hash())?
+            .iter()
+            .position(|key| {
+                let header = self
+                    .map
+                    .get(key)
+                    .expect("keys stored in the reverse index must exist in the map");
+                verify_rollup_blob_against_sequencer_blob(rollup, header)
+            })?;
```

Comment:
this returns the index of the first matching key, right? would there ever be a case where there are multiple keys in the `reverse_index.get(&rollup.sequencer_block_hash())` `Vec<Key>` that match the rollup blob?

Reply:
I commented above: this code looks like its defending against something that should be caught during the `verify` phase.

Reply:
The verify phase is now constructing a map `block_hash -> CelestiaSequencerBlob` so that I could simplify this code.